File: go/indicators/jeanphilippepoton/fractaladaptivesimplemovingaverage2/fractaladaptivesimplemovingaverage2.go
package fractaladaptivesimplemovingaverage2

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
)

// FractalAdaptiveSimpleMovingAverage2 computes the FRASMA2 indicator.
//
// Uses the corrected Fractal Graph Dimension Index (FGDI) formula to adaptively modify
// an SMA's period. The corrected formula uses ln(2*(N-1)) denominator and N-1
// path segments. When the market is trending (FGDI near 1.0), the SMA speeds up;
// when erratic (FGDI near 2.0), the SMA slows down.
//
// The indicator is not primed during the first `period` updates.
type FractalAdaptiveSimpleMovingAverage2 struct {
	mu sync.RWMutex
	core.LineIndicator
	window      []float64
	closes      []float64
	period      int
	normalSpeed int
	windowCount int
	primed      bool
	log2Pm1     float64
	ln2         float64
	invPSq      float64
}

// NewFractalAdaptiveSimpleMovingAverage2 returns an instance of the indicator created using supplied parameters.
func NewFractalAdaptiveSimpleMovingAverage2(p *Params) (*FractalAdaptiveSimpleMovingAverage2, error) {
	const (
		invalid = "invalid fractal adaptive simple moving average 2 parameters"
		fmts    = "%s: %s"
		fmtw    = "%s: %w"
		fmtn    = "frasma2(%d,%d%s)"
	)

	period := p.Period
	if period < 2 {
		return nil, fmt.Errorf(fmts, invalid, "period should be greater than 1")
	}

	normalSpeed := p.NormalSpeed
	if normalSpeed < 1 {
		return nil, fmt.Errorf(fmts, invalid, "normal_speed should be greater than 0")
	}

	bc := p.BarComponent
	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	qc := p.QuoteComponent
	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	tc := p.TradeComponent
	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	mnemonic := fmt.Sprintf(fmtn, period, normalSpeed, core.ComponentTripleMnemonic(bc, qc, tc))
	desc := "Fractal adaptive simple moving average 2 " + mnemonic

	f := &FractalAdaptiveSimpleMovingAverage2{
		window:      make([]float64, period),
		closes:      make([]float64, 0, 256),
		period:      period,
		normalSpeed: normalSpeed,
		log2Pm1:     math.Log(2.0 * float64(period-1)),
		ln2:         math.Log(2.0),
		invPSq:      1.0 / float64(period*period),
	}

	f.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, f.Update)

	return f, nil
}

// IsPrimed indicates whether the indicator is primed.
func (f *FractalAdaptiveSimpleMovingAverage2) IsPrimed() bool {
	f.mu.RLock()
	defer f.mu.RUnlock()

	return f.primed
}

// Metadata describes the output data of the indicator.
func (f *FractalAdaptiveSimpleMovingAverage2) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.FractalAdaptiveSimpleMovingAverage2,
		f.LineIndicator.Mnemonic,
		f.LineIndicator.Description,
		[]core.OutputText{
			{Mnemonic: f.LineIndicator.Mnemonic, Description: f.LineIndicator.Description},
		},
	)
}

// Update updates the value of the indicator given the next sample.
func (f *FractalAdaptiveSimpleMovingAverage2) Update(sample float64) float64 {
	if math.IsNaN(sample) {
		return sample
	}

	f.mu.Lock()
	defer f.mu.Unlock()

	period := f.period

	// Accumulate close history for SMA computation.
	f.closes = append(f.closes, sample)

	// Fill the FGDI window. FRASMA2 requires `period+1` samples before first output
	// (first `period` values are NaN, output starts at index `period`).
	if f.windowCount < period {
		f.window[f.windowCount] = sample
		f.windowCount++

		if f.windowCount < period {
			return math.NaN()
		}

		// windowCount == period: we have exactly `period` samples (indices 0..period-1).
		// FRASMA2 outputs NaN for the first `period` samples, so return NaN here
		// and mark primed. The next call will produce the first valid output.
		f.primed = true

		return math.NaN()
	}

	// Shift window.
	for i := 0; i < period-1; i++ {
		f.window[i] = f.window[i+1]
	}

	f.window[period-1] = sample

	// --- Compute FGDI using corrected formula (N-1 segments, ln(2*(N-1)) denominator) ---
	priceMax := f.window[0]
	priceMin := f.window[0]

	for k := 1; k < period; k++ {
		if f.window[k] > priceMax {
			priceMax = f.window[k]
		}

		if f.window[k] < priceMin {
			priceMin = f.window[k]
		}
	}

	priceRange := priceMax - priceMin
	if priceRange <= 0.0 {
		return math.NaN()
	}

	// N-1 segments: iterate from index 1 to period-1.
	priorNorm := (f.window[0] - priceMin) / priceRange
	length := 0.0

	for k := 1; k < period; k++ {
		currNorm := (f.window[k] - priceMin) / priceRange
		diff := currNorm - priorNorm
		length += math.Sqrt(diff*diff + f.invPSq)
		priorNorm = currNorm
	}

	if length <= 0.0 {
		return math.NaN()
	}

	fgdi := 1.0 + (math.Log(length)+f.ln2)/f.log2Pm1

	// --- Adaptive speed ---
	denom := 2.0 - fgdi
	if math.Abs(denom) < 1e-10 {
		return math.NaN()
	}

	trailDim := 1.0 / denom
	alpha := trailDim / 2.0
	speed := int(math.Round(float64(f.normalSpeed) * alpha))

	if speed < 1 {
		speed = 1
	}

	// --- SMA of length `speed` ending at current position ---
	nCloses := len(f.closes)
	if speed > nCloses {
		return math.NaN()
	}

	smaSum := 0.0
	for k := nCloses - speed; k < nCloses; k++ {
		smaSum += f.closes[k]
	}

	return smaSum / float64(speed)
}

File: py/indicators/jean_philippe_poton/fractal_adaptive_simple_moving_average_2/fractal_adaptive_simple_moving_average_2.py
"""Fractal Adaptive Simple Moving Average 2 (FRASMA2) indicator."""

import math

from ...core.indicator import Indicator
from ...core.line_indicator import LineIndicator
from ...core.metadata import Metadata
from ...core.build_metadata import build_metadata, OutputText
from ...core.identifier import Identifier
from ...core.component_triple_mnemonic import component_triple_mnemonic
from ...core.output import Output
from ....entities.bar import Bar
from ....entities.quote import Quote
from ....entities.trade import Trade
from ....entities.scalar import Scalar
from ....entities.bar_component import BarComponent, DEFAULT_BAR_COMPONENT, bar_component_value
from ....entities.quote_component import QuoteComponent, DEFAULT_QUOTE_COMPONENT, quote_component_value
from ....entities.trade_component import TradeComponent, DEFAULT_TRADE_COMPONENT, trade_component_value
from .params import FractalAdaptiveSimpleMovingAverage2Params


class FractalAdaptiveSimpleMovingAverage2(Indicator):
    """Computes the Fractal Adaptive Simple Moving Average 2 (FRASMA2).

    Uses the corrected Fractal Graph Dimension Index (FGDI) formula to adaptively modify
    an SMA's period. The corrected formula uses ln(2*(N-1)) denominator and N-1
    path segments. When the market is trending (FGDI near 1.0), the SMA speeds up;
    when erratic (FGDI near 2.0), the SMA slows down.

    The indicator is not primed during the first `period` updates.
    """

    def __init__(self, params: FractalAdaptiveSimpleMovingAverage2Params) -> None:
        period = params.period
        normal_speed = params.normal_speed

        if period < 2:
            raise ValueError(
                "invalid fractal adaptive simple moving average 2 parameters: period should be greater than 1")
        if normal_speed < 1:
            raise ValueError(
                "invalid fractal adaptive simple moving average 2 parameters: normal_speed should be greater than 0")

        bc = params.bar_component if params.bar_component is not None else DEFAULT_BAR_COMPONENT
        qc = params.quote_component if params.quote_component is not None else DEFAULT_QUOTE_COMPONENT
        tc = params.trade_component if params.trade_component is not None else DEFAULT_TRADE_COMPONENT

        bar_func = bar_component_value(bc)
        quote_func = quote_component_value(qc)
        trade_func = trade_component_value(tc)

        mnemonic = f"frasma2({period},{normal_speed}{component_triple_mnemonic(bc, qc, tc)})"
        description = f"Fractal adaptive simple moving average 2 {mnemonic}"

        self._line = LineIndicator(mnemonic, description, bar_func, quote_func, trade_func, self.update)

        self._period: int = period
        self._normal_speed: int = normal_speed
        self._window: list[float] = [0.0] * period
        self._window_count: int = 0
        self._closes: list[float] = []
        self._primed: bool = False

        # Precompute constants for corrected FGDI.
        self._log_2pm1: float = math.log(2.0 * (period - 1))
        self._ln2: float = math.log(2.0)
        self._inv_p_sq: float = 1.0 / (period * period)

    def is_primed(self) -> bool:
        return self._primed

    def metadata(self) -> Metadata:
        return build_metadata(
            Identifier.FRACTAL_ADAPTIVE_SIMPLE_MOVING_AVERAGE_2,
            self._line.mnemonic,
            self._line.description,
            [OutputText(self._line.mnemonic, self._line.description)],
        )

    def update(self, sample: float) -> float:
        if math.isnan(sample):
            return sample

        period = self._period

        # Accumulate close history for SMA computation.
        self._closes.append(sample)

        # Fill the FGDI window. First `period` values are NaN.
        if self._window_count < period:
            self._window[self._window_count] = sample
            self._window_count += 1

            if self._window_count < period:
                return math.nan

            self._primed = True

            return math.nan
        else:
            # Shift window left by one.
            for i in range(period - 1):
                self._window[i] = self._window[i + 1]
            self._window[period - 1] = sample

        # --- Compute FGDI using corrected formula (N-1 segments) ---
        price_max = self._window[0]
        price_min = self._window[0]
        for k in range(1, period):
            if self._window[k] > price_max:
                price_max = self._window[k]
            if self._window[k] < price_min:
                price_min = self._window[k]

        price_range = price_max - price_min

        if price_range <= 0.0:
            return math.nan

        # N-1 segments: iterate from index 1 to period-1.
        prior_norm = (self._window[0] - price_min) / price_range
        length = 0.0

        for k in range(1, period):
            curr_norm = (self._window[k] - price_min) / price_range
            diff = curr_norm - prior_norm
            length += math.sqrt(diff * diff + self._inv_p_sq)
            prior_norm = curr_norm

        if length <= 0.0:
            return math.nan

        fgdi = 1.0 + (math.log(length) + self._ln2) / self._log_2pm1

        # --- Adaptive speed ---
        denom = 2.0 - fgdi
        if abs(denom) < 1e-10:
            return math.nan

        trail_dim = 1.0 / denom
        alpha = trail_dim / 2.0
        speed = max(1, int(round(self._normal_speed * alpha)))

        # --- SMA of length `speed` ending at current position ---
        n_closes = len(self._closes)
        if speed > n_closes:
            return math.nan

        sma_sum = 0.0
        for k in range(n_closes - speed, n_closes):
            sma_sum += self._closes[k]

        return sma_sum / speed

    def update_scalar(self, sample: Scalar) -> Output:
        return self._line.update_scalar(sample)

    def update_bar(self, sample: Bar) -> Output:
        return self._line.update_bar(sample)

    def update_quote(self, sample: Quote) -> Output:
        return self._line.update_quote(sample)

    def update_trade(self, sample: Trade) -> Output:
        return self._line.update_trade(sample)
